### ISSUE TITLE
Bad performance when using Instancio for Entities in a data model

### ISSUE THREAD

**Issue by jur-clerkx** [instancio/instancio#391]
**Describe the bug**
We have a somewhat big data model, that has references between all the classes. I'm trying to create an object, that has a single reference to another object. This reference field is being set to a custom value using the `set` method on `InstancioOfClassAPI`. I suspect that instancio creates a builder, that also is able to build this reference class, which also has references to other classes. If this is the case, then this ends up in Instancio making a builder for the whole data model (+- 30 classes). This ends up in the `Instancio.of(<myclass>)` call taking up to around 1 second, making it not usable for my tests. (Also tried this on a DTO without references to other classes, and here it ran within a few milliseconds as expected.)

**Expected result**
Ideally, if a field is ignored or set, the api won't create a builder that includes this field, so that performance can be improved a lot.

**Actual result**
Seems like now, Instancio always creates a builder that is possible to populate all the fields, making it slow on "bigger" data models.

**Additional information**
Add any other context about the problem here

**Sample code reproducing the issue**
```java
Instancio.of<myclass>.class).set(field("myfieldToOtherClass"), customCreatedInstance).create()
```

- Instancio version:
2.6.0

**Comment by armandino** [comment]
@jur-clerkx, thank you for reporting this. In general, Instancio does not provide any performance guarantees, as long as the performance can be considered reasonable for a unit test. It's hard to quantify _reasonable_ because that obviously depends on the amount of generated data and complexity of the data model.

Regarding the InstancioOfClassApi builder object, there's only one instance created per `Instancio.create` and `Instancio.of`. This is also true when generating collections and streams.

If you could provide a reproducible sample, that would help a lot to troubleshoot (proprietary classes can be renamed, if that's a concern).  If not, then I can think of the following workarounds to speed things up:

- set default collection sizes to [min=1, max=2] via [Settings](https://www.instancio.org/user-guide/#overriding-settings-programmatically)
- `ignore()` fields that are not needed by a particular test case (especially, cyclic `OneToMany` collection fields).

I suspect that your class probably has many cyclic relationships. Instancio will try to generate data until a cycle is detected, which is then terminated with a null value. A sample would be helpful to investigate further.

**Comment by jur-clerkx** [comment]
@armandino Thanks for the reply. Sadly I can't provide the code, since this is the code of a client of ours.

As you suggested, I've tried to use the `ignore()` method for relation we don't use. However, this didn't change anything in the performance of the creation of the object (still, 3 objects takes around 5 seconds), although the generated object indeed did not have the external field filled.

I ran the test in a profiler, and it seems that the issues is in the `NodeFactory. createChildrenFromFields` method. It will still create nodes for the ignored field, which in turn will cause for the whole whole data model to be processed.

**Comment by armandino** [comment]
@jur-clerkx can you try the latest version `2.7.0` and see if you notice any difference?

**Comment by armandino** [comment]
@jur-clerkx I'm going to close this issue. Please re-open if you're still having issues after upgrading to `2.7.0`.

**Comment by jur-clerkx** [comment]
Hi @armandino,

I've retested this, with the `2.8.0` version. Performance seems to have improved slightly (3,5 seconds to generate 3 objects), but still too high to actually use.

Some context for our project. We've around 20-25 classes which are connected to each other. Most of these are 1-* relations. Using `ignore()` or a `Settings` object to reduce the collection size doesn't help here. 

As you can see in the Flame Graph below from the profiler, almost all of the times is in creating the NodeFactory to create the objects. Seems to me that https://github.com/instancio/instancio/issues/428 is also related to how this part of the library functions. I think that including the selectors (e.g. `set` or `ignore`) in the first stage, because if these are used, you wouldn't need the Nodes for these fields.

![image](https://user-images.githubusercontent.com/8639292/218721851-a9395345-a509-48b5-8401-58b459938632.png)

**Comment by armandino** [comment]
@jur-clerkx thanks for the follow up. It's good to hear there was at least a minor improvement. I'd love to investigate this further, but I will need a reproducible sample to see what's going on.

If you can provide a sample with the classes renamed, that would help a lot. Or if you prefer to provide it privately, feel free to reach me directly at my GH username `@` gmail.

**Comment by armandino** [comment]
@jur-clerkx not sure if you had a chance to try version `2.9.0`. I believe it should improve the performance issue you were having. At least you should be able to speed things up now by setting `withMaxDepth()` to a lower number (currently, the default is 8), e.g.:

``` java
Foo foo = Instancio.of(Foo.class)
    .withMaxDepth(4)
    .create();
```

You also made a good point about handling `ignore()` during node creation (the first stage). I made changes to the implementation based on your suggestion in #439. Now using `ignore()` should actually speed things up. This change will be released in version `2.10.0`.

I'll keep looking for other areas to make improvements. In the meantime, let me know if you run into any other issues. Thank you again for looking into this!
